lib/ssl/src/tls_connection.erl

```diff
@@ -669,21 +670,24 @@ hello(internal, #client_hello{extensions = Extensions} = Hello,
                                          handshake_env = HsEnv#handshake_env{hello = Hello}},
      [{reply, From, {ok, Extensions}}]};
 hello(internal, #server_hello{extensions = Extensions} = Hello,
-      #state{ssl_options = #{handshake := hello},
+      #state{ssl_options = #{
+                 handshake := hello,
+                 ocsp_stapling := OcspStapling},
              handshake_env = HsEnv,
              start_or_recv_from = From} = State) ->
     %% RFC6066.8, If a server returns a "CertificateStatus" message,
     %% then the server MUST have included an extension of type
     %% "status_request" with empty "extension_data" in the extended
     %% server hello.
     OcspState = HsEnv#handshake_env.ocsp_stapling_state,
-    IsCertStatusRequested = maps:is_key(ocsp_stapling, Extensions),
+    IsCertStatusRequested = (undefined == maps:get(status_request, Extensions, false)),
```

Comment:
At this point we have received the `server_hello` and we had sent out an OCSPRequest in ClientHello based on the value of the `ocsp_stapling` option. Next, we should check the status_request extension:
- if `ocsp_stapling`  is set to `false`, that is the client never sent out an OCSPRequest, then the presence of the status_request extension is a protocol error and we should abort the handshake. 
- if `ocsp_stapling` is set to `true`, then we should check if the `status_request` is empty (undefined). If `status_request` is empty we have successfully negotiated OCSP and we can set the `IsCertStatusRequested` (should be called `OcspNegotiated`) to true. If `status_request` not empty that it is a protocol error and we should abort the handshake.
I think this logic should go into a separate function with 2 input parameters: `OcspStapling` and `Extensions`:
```erlang
OcspNegotiated = handle_status_request_extension(OcspStapling, Extenions),
...
handle_status_request_extension(false = OcspStapling, Extensions) ->
    case maps:get(status_request, Extensions, false) of
        false -> %% extension is not present
            OcspStapling;
        _Else -> %% protocol error
           %% error handling here
    end;
handle_status_request_extension(true = OcspStapling, Extensions) ->
    case maps:get(status_request, Extensions, false) of
        undefined -> %% OCSP negotiated
            OcspStapling;
        false -> %% server won't negotiate OCSP
            false;
        _Else -> %% protocol error, status_request not empty
           %% error handling here
    end.
```
@IngelaAndin